### PR TITLE
Ensure compliant node construction/destruction API.

### PR DESCRIPTION
**Pull request by hidmic**
Connected to https://github.com/ros2/rmw/pull/249.

### CODE REVIEW

**Comment by hidmic** [comment]
Conflicts resolved. Also adjusted changes based on discussions in https://github.com/ros2/rmw/pull/249 and connected PRs.

@eboasson PTAL!

**Comment by hidmic** [comment]
One last rebase, I made a mistake in the previous one.

**Comment by hidmic** [comment]
One last rebase, I made a mistake in the previous one.

**Comment by hidmic** [comment]
Running CI up to `test_rmw_implementation` and all RMW implementations:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11444/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6659/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9388/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11347/)

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 1]
```diff
@@ -1310,11 +1340,14 @@ extern "C" rmw_ret_t rmw_destroy_node(rmw_node_t * node)
       common->pub, static_cast<void *>(&participant_msg), nullptr);
   }
 
-  rmw_free(const_cast<char *>(node->name));
-  rmw_free(const_cast<char *>(node->namespace_));
-  node->context->impl->fini();
-  rmw_node_free(node);
+  assert(node->context != nullptr);
```
Comment (clalancette) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:1343]: As with the other PRs around this, I'll suggest that we elide this check completely.  See the conversation at https://github.com/ros2/rmw_fastrtps/pull/408#discussion_r454341097 for more details on to why.

Comment (hidmic) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Uh, I thought I had removed them all. There, see https://github.com/ros2/rmw_cyclonedds/pull/206/commits/49992cdf1ce22c9cf25f81262444eae0f6443362

**Pull request merged**